pkg/clusteragent/custommetrics/store.go

```diff
@@ -0,0 +1,136 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+// +build kubeapiserver
+
+package custommetrics
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// Store is an interface for persistent storage of custom and external metrics.
+type Store interface {
+	SetExternalMetric(ExternalMetricValue) error
+	DeleteExternalMetric(hpaNamespace, hpaName, metricName string) error
+	ListAllExternalMetrics() ([]ExternalMetricValue, error)
+	Update() error
+}
+
+// configMapStore provides persistent storage of custom and external metrics using a configmap.
+type configMapStore struct {
+	namespace string
+	name      string
+	client    corev1.CoreV1Interface
+	cm        *v1.ConfigMap
+}
+
+// NewConfigMapStore returns a new store backed by a configmap. The configmap will be created
+// in the specified namespace if it does not exist.
+func NewConfigMapStore(client corev1.CoreV1Interface, ns, name string) (Store, error) {
+	cm, err := client.ConfigMaps(ns).Get(name, metav1.GetOptions{})
+	if err == nil {
+		log.Infof("Retrieved the configmap %s", name)
+		return &configMapStore{
+			namespace: ns,
+			name:      name,
+			client:    client,
+			cm:        cm,
+		}, nil
+	}
+
+	if !errors.IsNotFound(err) {
+		log.Infof("Error while attempting to fetch the configmap %s: %s", name, err)
+		return nil, err
+	}
+
+	log.Infof("The configmap %s not dot exist, trying to create it", name)
+	cm = &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+		},
+	}
+	// FIXME: distinguish RBAC error
+	cm, err = client.ConfigMaps(ns).Create(cm)
+	if err != nil {
+		return nil, err
+	}
+	return &configMapStore{
+		namespace: ns,
+		name:      name,
+		client:    client,
+		cm:        cm,
+	}, nil
+}
+
+// SetExternalMetric updates the associated external metric in the cached configmap.
+func (c *configMapStore) SetExternalMetric(em ExternalMetricValue) error {
```

Comment:
This updates the local copy of the configmap. The client is expected to call `Update` to actually persist the changes.